executor/window_parallel.go

```diff
@@ -0,0 +1,367 @@
+	//e.concurrency = e.ctx.GetSessionVars().WindowConcurrency
+
+	e.finishCh = make(chan struct{}, 1)
+	e.outputCh = make(chan *windowParallelOutput, e.concurrency)
+
+	e.workers = make([]windowParallelWorker, e.concurrency)
+	for i := range e.workers {
+		e.workers[i] = windowParallelWorker{
+			ctx:      e.ctx,
+			e:        e,
+			finishCh: e.finishCh,
+
+			inputCh:        make(chan *chunk.Chunk, 1),
+			inputHolderCh:  make(chan *chunk.Chunk, 1),
+			outputCh:       e.outputCh,
+			outputHolderCh: make(chan *chunk.Chunk, 1),
+
+			dataFetcher: e.dataFetchers[i],
+			sortExec:    e.sortExecs[i],
+			windowExec:  e.windowExecs[i],
+		}
+		e.workers[i].dataFetcher.finishCh = e.finishCh
+		e.workers[i].dataFetcher.inputCh = e.workers[i].inputCh
+		e.workers[i].dataFetcher.giveBackCh = e.workers[i].inputHolderCh
+		if err := e.workers[i].windowExec.Open(ctx); err != nil {
+			return err
+		}
+
+		e.workers[i].inputHolderCh <- newFirstChunk(e.children[0])
+		e.workers[i].outputHolderCh <- newFirstChunk(e)
+	}
+
+	return nil
+}
+
+// Close implements the Executor Close interface
+func (e *WindowParallelExec) Close() error {
+	if !e.prepared {
+		for i := range e.workers {
+			close(e.workers[i].inputHolderCh)
+			close(e.workers[i].inputCh)
+			close(e.workers[i].outputHolderCh)
+		}
+		close(e.outputCh)
+	}
+	close(e.finishCh)
+	for i := range e.workers {
+		for range e.workers[i].inputCh {
+		}
+	}
+	for range e.outputCh {
+	}
+	e.executed = false
+
+	if e.runtimeStats != nil {
+		e.runtimeStats.SetConcurrencyInfo("WindowParallelConcurrency", e.concurrency)
+	}
+
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+func recoveryWindowParallelExec(output chan *windowParallelOutput, r interface{}) {
+	err := errors.Errorf("%v", r)
+	output <- &windowParallelOutput{err: errors.Errorf("%v", r)}
+	logutil.BgLogger().Error("window parallel panicked", zap.Error(err))
+}
+
+func (e *WindowParallelExec) execDataFetcherThread(ctx context.Context) {
+	var (
+		ok        bool
+		err       error
+		groupKeys [][]byte
+	)
+	results := make([]*chunk.Chunk, len(e.workers))
+
+	defer func() {
+		if r := recover(); r != nil {
+			recoveryWindowParallelExec(e.outputCh, r)
+		}
+		for i := range e.workers {
+			close(e.workers[i].inputCh)
+		}
+	}()
+	for {
+		chk := newFirstChunk(e.children[0])
+		err = Next(ctx, e.children[0], chk)
+		if err != nil {
+			e.outputCh <- &windowParallelOutput{err: err}
+			return
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		groupKeys, err = getGroupKey(e.ctx, chk, groupKeys, e.groupByItems)
+		if err != nil {
+			e.outputCh <- &windowParallelOutput{err: err}
+			return
+		}
+		numRows := chk.NumRows()
+		for i := 0; i < numRows; i++ {
+			workerIdx := int(murmur3.Sum32(groupKeys[i])) % len(e.workers)
+			w := &e.workers[workerIdx]
+
+			if results[workerIdx] == nil {
+				select {
+				case <-e.finishCh:
+					return
+				case results[workerIdx], ok = <-w.inputHolderCh:
```

Comment:
When will `ok` be false?

Reply:
En...no matter when, I don't close the channel.
I just think that more correct than necessary, would lead to be more robust.
So it is better to ignore `ok` ?